mqtt_mode.go

```diff
@@ -0,0 +1,250 @@
+// Source for the MODE specific MQTT protocol.
+package mode_client
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+)
+
+type (
+	mqttMsgHandler func([]byte) error
+
+	// DeviceCommand represents a received from the MODE cloud.
+	DeviceCommand struct {
+		Action  string
+		payload json.RawMessage
+	}
+
+	KeyValueSync struct {
+		Action   string      `json:"action"`
+		Revision int         `json:"rev"`
+		Key      string      `json:"key"`
+		Value    interface{} `json:"value"`
+		NumItems int         `json:"numItems"`
+		Items    []*KeyValue `json:"items"`
+	}
+
+	// KeyValue represents a key-value pair stored in the Device Data Proxy.
+	KeyValue struct {
+		Key              string      `json:"key"`
+		Value            interface{} `json:"value"`
+		ModificationTime time.Time   `json:"modificationTime"`
+	}
+
+	// DeviceEvent represents an event to be sent to the MODE cloud.
+	DeviceEvent struct {
+		EventType string                 `json:"eventType"`
+		EventData map[string]interface{} `json:"eventData,omitempty"`
+		Qos       QOSLevel               // not exported to JSON
+	}
+
+	// BulkData represents a batch of opaque data to be sent to the MODE cloud.
+	DeviceBulkData struct {
+		StreamID string
+		Blob     []byte
+		Qos      QOSLevel // not exported to serializer
+	}
+
+	// Implements the MqttDelegate
+	ModeMqttDelegate struct {
```

Comment:
`ModeMqttDelegate` can be seen from the side of MqttClient?
Normally it leaves MQTT communication to MqttClient instance?

Reply:
MqttClient will call methods on MqttDelegate, but it doesn't know about ModeMqttDelegate. Is that your question?

Reply:
Yes! Thank you!